OSSM-4565: [DOC] Add Kiali Configuration for user-workload monitoring (#62902)
openshift/openshift-docs: modules/ossm-integrating-with-user-workload-monitoring.adoc

* User-workload monitoring is enabled.
* {SMProductName} Operator 2.4 is installed.
* Kiali Operator 1.65 is installed.

.Procedure

> . Configure Kiali for `user-workload-monitoring`:

Reviewer: ```suggestion
. Configure Kiali for `user-workload monitoring`:
```

+
[source,yaml]
----
apiVersion: kiali.io/v1alpha1
kind: Kiali
metadata:
  name: kiali-user-workload-monitoring

>   namespace: istio-system-1

Reviewer: ```suggestion
  namespace: istio-system
```

spec:
  external_services:
    istio:
      config_map_name: istio-basic
      istio_sidecar_injector_config_map_name: istio-sidecar-injector-basic
      istiod_deployment_name: istiod-basic
      istiod_pod_monitoring_port: 15014

>       url_service_version: 'http://istiod-basic.istio-system-1:15014/version'

Reviewer: ```suggestion
      url_service_version: 'http://istiod-basic.istio-system:15014/version'
```

    prometheus:
      auth:
        token: secret:thanos-querier-web-token:token
        type: bearer
        use_kiali_token: false
      query_scope:

>         mesh_id: "basic_istio-system-1"

Reviewer: ```suggestion
        mesh_id: "basic_istio-system"
```
Btw. now I see that we should change `replacement: "<smcp_name>-<smcp_namespace>" # <2>` in Pod/Service monitors and explain in the <2> that this value could be anything unique, e.g. `<smcp_name>-<smcp_namespace>`.

* User-workload monitoring is enabled.
* {SMProductName} Operator 2.4 is installed.
* Kiali Operator 1.65 is installed.

.Procedure

. Create a token to Thanos for Kiali:
+
[source,terminal]
----
$ SECRET=`oc get secret -n openshift-user-workload-monitoring | 
 grep  prometheus-user-workload-token | head -n 1 | awk '{print $1 }'`
----
+
[source,terminal]
----

> $ TOKEN=`echo $(oc get secret $SECRET -n openshift-user-workload-monitoring -o jsonpath='{.data.token}' | base64 -d`

Reviewer: ```suggestion
$ TOKEN=`echo $(oc get secret $SECRET -n openshift-user-workload-monitoring -o jsonpath='{.data.token}') | base64 -d`
```
It would be great if we could remove `$`, because it does not allow to copy the line and paste in the terminal as is. Instead, we have to copy and remove first character. This is so annoying... Is it necessary?

Author: Yes, unfortunately, the `$` has to remain per Red Hat style guidelines:

https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#code-blocks-command-syntax-and-example-output

If the user must run a command as root, use a number sign (#) at the start of the command instead of a dollar sign ($). For example:

# subscription-manager list

Reviewer 2: @jewertow and @gwynnemonahan - FYI, when a user clicks the Copy icon next to sample code in the docs, the prompt is not included in the copied text:

![image](https://github.com/openshift/openshift-docs/assets/86380613/6d3e1967-a1bf-4c85-ba3b-a9a85bf78f9b)

----
+
[source,terminal]
----
$ oc create secret generic thanos-querier-web-token -n istio-system --from-literal=token=$TOKEN
----

. Configure Kiali for user-workload monitoring:
+
[source,yaml]
----
apiVersion: kiali.io/v1alpha1
kind: Kiali
metadata:
  name: kiali-user-workload-monitoring
  namespace: istio-system
spec:
  external_services:
    istio:
      config_map_name: istio-basic
      istio_sidecar_injector_config_map_name: istio-sidecar-injector-basic
      istiod_deployment_name: istiod-basic
      istiod_pod_monitoring_port: 15014
      url_service_version: 'http://istiod-basic.istio-system:15014/version'
    prometheus:
      auth:
        token: secret:thanos-querier-web-token:token
        type: bearer
        use_kiali_token: false
      query_scope:

>         mesh_id: "basic_istio-system"

Reviewer: It must match values from PodMonitor:
```suggestion
        mesh_id: "basic-istio-system"
```

    interval: 30s
    relabelings:
    - action: replace
      replacement: "basic-istio-system" # <2>
      targetLabel: mesh_id # <3>
----
<1> Since {product-title} monitoring ignores the `namespaceSelector` spec in `ServiceMonitor` and `PodMonitor` objects, you must apply the `PodMonitor` object in all mesh namespaces, including the control plane namespace.

> <2> Since `"basic-istio-system"` is a combination of the SMCP name and namespace, you can replace it with any other unique value within a cluster with multiple SMCPs. You must replace `"basic-istio-system"` if you are using Kiali and have multiple SMCPs deployed in a cluster. You do not need to do this if you have a cluster-wide deployment. 

Reviewer: 1) "You must replace `"basic-istio-system"` if you are using Kiali and have multiple SMCPs deployed in a cluster." - Replace with what? I would prefer to just say that the value of `mesh_id` label must be unique for every mesh in the cluster.
2) "You do not need to do this if you have a cluster-wide deployment." - this sentence is very unclear. What don't I have to do? Changing "basic-istio-system" or adding it at all? I wanted to explain that `mesh_id` label is not necessary in cluster-wide deployments and then users don't need it in Pod/ServiceMonitor as well as in Kiali. In cluster-wide deployments users can remove this replacement:
```
    - action: replace
      replacement: "<smcp_name>-<smcp_namespace>" # <2>
      targetLabel: mesh_id
```
We can explain 2. in <3>.

    interval: 30s
    relabelings:
    - action: replace
      replacement: "basic-istio-system" # <2>
      targetLabel: mesh_id
----
<1> Since {product-title} monitoring ignores the `namespaceSelector` spec in `ServiceMonitor` and `PodMonitor` objects, you must apply the `PodMonitor` object in all mesh namespaces, including the control plane namespace.

> <2> The variable `"basic-istio-system"` is a combination of the SMCP name and namespace. It can be replaced with any other unique value within a cluster with multiple SMCPs. This is required only when Kiali is used and multiple SMCPs are deployed in a cluster. In case of cluster-wide deployment, it can be skipped. A shared Prometheus instance can include metrics from multiple meshes. However, if you are deploying Kiali, the metrics must be associated with a single mesh. To associate metrics with a single mesh, add a unique `mesh_id` to each one. Doing so narrows the query scope in Kiali to only the relevant mesh metrics.

Reviewer 3: ```suggestion
<2> Adding the `mesh_id` label can be skipped if there will never be more than one mesh being used with user workload monitoring on the cluster (be sure to also remove the `spec.prometheus.query_scope` from the Kiali deployment if following this guide). Otherwise, adding a `mesh_id` label is recommended as it allows for scoping metrics to their mesh of origin, and for Kiali it is required. The string `"basic-istio-system"` is a combination of the SMCP name and its namespace, but any label can be used as long as it is unique for every mesh using user workload monitoring in the cluster.
```

I took a stab at re-wording this to fix the following issues:

- For clarity, I think `"basic-istio-system"` should be called a "string" not a "variable".
- Additionally, the connection is never made between this value and the `spec.prometheus.query_scope` of the Kiali example resource given above, so it gives the impression that if I only have one mesh in my cluster I can just leave out this `mesh_id` relabeling, when I would also have to clear the `spec.prometheus.query_scope` of the Kiali resource.
- It also misses the nuance that the issue doesn't come from having more than one mesh _on the cluster_, in and of itself. There is only a need for the `mesh_id` label when there is more than one mesh being _who's metrics are being collected by user workload monitoring stack_. If including this detail makes the footnote too long or hard to follow, I don't think it's an important enough distinction to justify itself.

Author: Taking your suggestions via Slack of a shorter footnote, and then including the other information as a NOTE. I think that will be easier to follow than cramming it all into a footnote, or trying to separate it all into a different module, esp since it is all part of the same step.

    interval: 30s
    relabelings:
    - action: replace
      replacement: "basic-istio-system" # <2>
      targetLabel: mesh_id
----
<1> Since {product-title} monitoring ignores the `namespaceSelector` spec in `ServiceMonitor` and `PodMonitor` objects, you must apply the `PodMonitor` object in all mesh namespaces, including the control plane namespace.
<2> The string `"basic-istio-system"` is a combination of the SMCP name and its namespace, but any label can be used as long as it is unique for every mesh using user workload monitoring in the cluster. The `spec.prometheus.query_scope` of the Kiali resource configured in Step 2 needs to match this value.
+
[NOTE]
====
If there is only one mesh using user-workload monitoring, then both the `mesh_id` relabeling and the `spec.prometheus.query_scope` field in the Kiali resource are optional (but the `query_scope` field given here should be removed if the `mesh_id` label is removed).

> If there may be multiple meshes using user-workload monitoring, then both the `mesh_id` relabelings and the `spec.prometheus.query_scope` field in the Kiali resource are required so that Kiali only sees metrics from its associated mesh. If Kiali is not being deployed, applying the `mesh_id` relabeling is still recommended so that metrics from different meshes can be distinguished from one another.

Reviewer 4: s/may/might per IBM SG

      action: replace
      targetLabel: pod_name
    - action: replace
      replacement: "basic-istio-system" # <2>
      targetLabel: mesh_id # <3>
----
<1> Since {product-title} monitoring ignores the `namespaceSelector` spec in `ServiceMonitor` and `PodMonitor` objects, you must apply the `PodMonitor` object in all mesh namespaces, including the control plane namespace.

> <2> Replace `"basic-istio-system"` with a unique value, such as `<smcp-name>-<smcp-namespace>` to create a unique combination within a cluster with multiple SMCPs. 

Reviewer: ```suggestion
<2> `"basic-istio-system"` is a combination of the SMCP name and namespace. It can be replaced with any other unique value within a cluster with multiple SMCPs. This is required only when Kiali is used and multiple SMCPs are deployed in a cluster. In case of cluster-wide deployment, it can be skipped.
```

      action: replace
      targetLabel: pod_name
    - action: replace

>       replacement: <smcp_name>-<smcp_namespace> # <2>

Author: Does "<smcp-name>-<smcp-namespace>" get replaced with "basic-istio-system" in both places? The first comment shows "basic-istio-system" but this comment you have it changing to `<smcp-name>-<smcp-namespace>` and not "basic-istio-system" so wanted to check.

>       replacement: <smcp_name>-<smcp_namespace> # <2>

Reviewer: > Does <smcp_name>-<smcp_namespace> get replaced with "basic-istio-system" in both places?

Yes.

      action: replace
      targetLabel: pod_name
    - action: replace
      replacement: "basic-istio-system" # <2>
      targetLabel: mesh_id
----
<1> Since {product-title} monitoring ignores the `namespaceSelector` spec in `ServiceMonitor` and `PodMonitor` objects, you must apply the `PodMonitor` object in all mesh namespaces, including the control plane namespace.
<2> The string `"basic-istio-system"` is a combination of the SMCP name and its namespace, but any label can be used as long as it is unique for every mesh using user workload monitoring in the cluster. The `spec.prometheus.query_scope` of the Kiali resource configured in Step 2 needs to match this value.
+
[NOTE]
====
If there is only one mesh using user-workload monitoring, then both the `mesh_id` relabeling and the `spec.prometheus.query_scope` field in the Kiali resource are optional (but the `query_scope` field given here should be removed if the `mesh_id` label is removed).

> If there may be multiple meshes using user-workload monitoring, then both the `mesh_id` relabelings and the `spec.prometheus.query_scope` field in the Kiali resource are required so that Kiali only sees metrics from its associated mesh. If Kiali is not being deployed, applying the `mesh_id` relabeling is still recommended so that metrics from different meshes can be distinguished from one another.

Reviewer 4: s/may/might
